File: src/eigen/jacobi.cpp
/*! 
  @file jacobi.cpp
	
  @brief 固有値・固有ベクトル
         ヤコビ(Jacobi)法
 
  @author Makoto Fujisawa
  @date 2012-06
*/


//-----------------------------------------------------------------------------
// インクルードファイル
//-----------------------------------------------------------------------------
#include "rx_utils.h"


/*!
 * Jacobi法による固有値の算出
 * @param[inout] a 実対称行列．計算後，対角要素に固有値が入る
 * @param[out] v 固有ベクトル(aと同じサイズ)
 * @param[in] n 行列のサイズ(n×n)
 * @param[inout] max_iter 最大反復数(反復終了後,実際の反復数を返す)
 * @param[inout] eps 許容誤差(反復終了後,実際の誤差を返す) 
 * @return 反復回数
 */
int eigen_jacobi(vector< vector<double> > &a, vector< vector<double> > &v, int n, int &max_iter, double &eps)
{
    vector<double> bim(n), bjm(n);
    double bii, bij, bjj, bji;
  
    for(int i = 0; i < n; ++i){
        for(int j = 0; j < n; ++j){
            v[i][j] = (i == j) ? 1.0 : 0.0;
        }
    }
 
	double e;
    int k;
    for(k = 0; k < max_iter; ++k){
        int i, j;
 
		// 非対角要素で絶対値が最大の要素をa_ijとして0にする
        double x = 0.0;
        for(int ia = 0; ia < n; ++ia){
            for(int ja = 0; ja < n; ++ja){
                if(ia != ja && fabs(a[ia][ja]) > x){
                    i = ia;
                    j = ja;
                    x = fabs(a[ia][ja]);
                }
            }
        }
 
		// sinθ,cosθを算出
        double aii = a[i][i];
        double ajj = a[j][j];
        double aij = a[i][j];
 
        double alpha, beta;
        alpha = (aii-ajj)/2.0;
        beta  = sqrt(alpha*alpha+aij*aij);
 
        double st, ct;
        ct = sqrt((1.0+fabs(alpha)/beta)/2.0);    // sinθ
        st = (((aii-ajj) >= 0.0) ? 1.0 : -1.0)*aij/(2.0*beta*ct);    // cosθ
 
        // A = PAPの計算
        for(int m = 0; m < n; ++m){
            if(m == i || m == j) continue;
 
            double aim = a[i][m];
            double ajm = a[j][m];
 
            bim[m] =  aim*ct+ajm*st;
            bjm[m] = -aim*st+ajm*ct;
        }
 
        bii = aii*ct*ct+2.0*aij*ct*st+ajj*st*st;
        bij = 0.0;
 
        bjj = aii*st*st-2.0*aij*ct*st+ajj*ct*ct;
        bji = 0.0;
 
        for(int m = 0; m < n; ++m){
            a[i][m] = a[m][i] = bim[m];
            a[j][m] = a[m][j] = bjm[m];
        }
        a[i][i] = bii;
        a[i][j] = bij;
        a[j][j] = bjj;
        a[j][i] = bji;
 
        // V = PVの計算
        for(int m = 0; m < n; ++m){
            double vmi = v[m][i];
            double vmj = v[m][j];
 
            bim[m] =  vmi*ct+vmj*st;
            bjm[m] = -vmi*st+vmj*ct;
        }
        for(int m = 0; m < n; ++m){
            v[m][i] = bim[m];
            v[m][j] = bjm[m];
        }
 
		// 非対角要素の絶対値の和で収束を判定
        e = 0.0;
        for(int ja = 0; ja < n; ++ja){
            for(int ia = 0; ia < n; ++ia){
                if(ia != ja){
                    e += fabs(a[ja][ia]);
                }
            }
        }
        if(e < eps){
			k++;
			break;
		}
    }

	max_iter = k;
	eps = e;

    return 1;
}



//-----------------------------------------------------------------------------
//! メイン関数
//-----------------------------------------------------------------------------
int main(void)
{
	// ファイルから行列要素を読み込んで解く
	vector< vector<double> > A;

	ReadMatrix("matrix.txt", ",", A);

	int n = (int)A.size();	// n×n行列
	cout << "A(" << n << " x " << n << ") = " << endl;
	OutputMatrix(A, n, n);
	cout << endl;

	vector< vector<double> > V(n, vector<double>(n, 0.0));

	int max_iter = 100;
	double eps = 1e-6;
	eigen_jacobi(A, V, n, max_iter, eps);

	// 固有値の表示
	cout << "e = (";
	for(int i = 0; i < n; ++i){
		cout << A[i][i] << (i == n-1 ? ")" : ", ");
	}
	cout << endl;

	// 固有ベクトルの表示
	for(int j = 0; j < n; ++j){
		cout << "v" << j << " = (";
		for(int i = 0; i < n; ++i){
			cout << V[i][j] << (i == n-1 ? ")" : ", ");
		}
		cout << endl;
	}
	cout << endl;

	cout << "iter = " << max_iter << ", eps = " << eps << endl;
	cout << endl;

	return 0;
}

File: python/eigen/jacobi.py
# -*- coding: utf-8 -*-
"""!
@file jacobi.py

@brief 固有値・固有ベクトル
       ヤコビ(Jacobi)法

@author Makoto Fujisawa
@date 2012-06 (Python版)
"""

# -----------------------------------------------------------------------------
# インポート
# -----------------------------------------------------------------------------
import os
import sys
import math

sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "shared"))
from rx_utils import *


def eigen_jacobi(a, v, n, max_iter, eps):
    """!
    Jacobi法による固有値の算出
    @param[inout] a 実対称行列．計算後，対角要素に固有値が入る
    @param[out] v 固有ベクトル(aと同じサイズ)
    @param[in] n 行列のサイズ(n×n)
    @param[in] max_iter 最大反復数
    @param[in] eps 許容誤差
    @return (戻り値, 実際の反復数, 実際の誤差)
    """
    bim = zeros(n)
    bjm = zeros(n)

    for i in range(n):
        for j in range(n):
            v[i][j] = 1.0 if i == j else 0.0

    e = 0.0
    k = 0
    for k in range(max_iter):
        # 非対角要素で絶対値が最大の要素をa_ijとして0にする
        i = 0
        j = 0
        x = 0.0
        for ia in range(n):
            for ja in range(n):
                if ia != ja and abs(a[ia][ja]) > x:
                    i = ia
                    j = ja
                    x = abs(a[ia][ja])

        # sinθ,cosθを算出
        aii = a[i][i]
        ajj = a[j][j]
        aij = a[i][j]

        alpha = (aii-ajj)/2.0
        beta = math.sqrt(alpha*alpha+aij*aij)

        ct = math.sqrt((1.0+abs(alpha)/beta)/2.0)                  # cosθ
        st = (1.0 if (aii-ajj) >= 0.0 else -1.0)*aij/(2.0*beta*ct)  # sinθ

        # A = PAPの計算
        for m in range(n):
            if m == i or m == j:
                continue

            aim = a[i][m]
            ajm = a[j][m]

            bim[m] = aim*ct+ajm*st
            bjm[m] = -aim*st+ajm*ct

        bii = aii*ct*ct+2.0*aij*ct*st+ajj*st*st
        bij = 0.0

        bjj = aii*st*st-2.0*aij*ct*st+ajj*ct*ct
        bji = 0.0

        for m in range(n):
            a[i][m] = a[m][i] = bim[m]
            a[j][m] = a[m][j] = bjm[m]
        a[i][i] = bii
        a[i][j] = bij
        a[j][j] = bjj
        a[j][i] = bji

        # V = PVの計算
        for m in range(n):
            vmi = v[m][i]
            vmj = v[m][j]

            bim[m] = vmi*ct+vmj*st
            bjm[m] = -vmi*st+vmj*ct
        for m in range(n):
            v[m][i] = bim[m]
            v[m][j] = bjm[m]

        # 非対角要素の絶対値の和で収束を判定
        e = 0.0
        for ja in range(n):
            for ia in range(n):
                if ia != ja:
                    e += abs(a[ja][ia])
        if e < eps:
            k += 1
            break

    return 1, k, e


# -----------------------------------------------------------------------------
# メイン関数
# -----------------------------------------------------------------------------
def main():
    # ファイルから行列要素を読み込んで解く
    fn = os.path.join(os.path.dirname(os.path.abspath(__file__)), "matrix.txt")
    A = ReadMatrix(fn, ",")

    n = len(A)  # n×n行列
    print("A(" + str(n) + " x " + str(n) + ") = ")
    OutputMatrix(A, n, n)
    print("")

    V = zeros2(n, n)

    max_iter = 100
    eps = 1e-6
    ret, max_iter, eps = eigen_jacobi(A, V, n, max_iter, eps)

    # 固有値の表示
    s = "e = ("
    for i in range(n):
        s += fmt(A[i][i]) + (")" if i == n-1 else ", ")
    print(s)

    # 固有ベクトルの表示
    for j in range(n):
        s = "v" + str(j) + " = ("
        for i in range(n):
            s += fmt(V[i][j]) + (")" if i == n-1 else ", ")
        print(s)
    print("")

    print("iter = " + str(max_iter) + ", eps = " + fmt(eps))
    print("")

    return 0


if __name__ == "__main__":
    main()
